File: src/lib/ai/tools/web-search.ts
// =============================================================================
// Tool: web_search — Tavily API (raw fetch, no SDK)
// =============================================================================

import type { AgentTool, ToolResult, AgentContext } from '@/lib/ai/types';

const TAVILY_URL = 'https://api.tavily.com/search';

async function handler(
  input: Record<string, unknown>,
  _ctx: AgentContext,
): Promise<ToolResult> {
  const apiKey = process.env.TAVILY_API_KEY;
  if (!apiKey) {
    return { content: 'Web search is not configured (TAVILY_API_KEY missing).', isError: true };
  }

  const query = String(input.query || '').trim();
  if (!query) {
    return { content: 'Missing required parameter: query', isError: true };
  }

  const maxResults = Math.min(Number(input.max_results) || 5, 10);
  const searchDepth = input.search_depth === 'basic' ? 'basic' : 'advanced';
  const includeAnswer = input.include_answer !== false;

  try {
    const res = await fetch(TAVILY_URL, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        api_key: apiKey,
        query,
        max_results: maxResults,
        search_depth: searchDepth,
        include_answer: includeAnswer,
        include_raw_content: false,
      }),
    });

    if (!res.ok) {
      const errText = await res.text().catch(() => 'Unknown error');
      return { content: `Tavily API error (${res.status}): ${errText}`, isError: true };
    }

    const data = await res.json();

    // Format results for the model
    let output = '';

    if (data.answer) {
      output += `**Summary:** ${data.answer}\n\n`;
    }

    if (data.results?.length) {
      output += '**Sources:**\n';
      for (const r of data.results) {
        output += `- [${r.title}](${r.url})\n  ${r.content?.slice(0, 300) || 'No snippet'}\n\n`;
      }
    } else {
      output += 'No results found for this query.';
    }

    return {
      content: output.trim(),
      metadata: {
        resultCount: data.results?.length || 0,
        query,
      },
    };
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    return { content: `Web search failed: ${msg}`, isError: true };
  }
}

export const webSearchTool: AgentTool = {
  name: 'web_search',
  description:
    'Search the web for current information, news, trends, or any external data. Use this when you need up-to-date information that is not in the knowledge base.',
  input_schema: {
    type: 'object',
    properties: {
      query: {
        type: 'string',
        description: 'The search query',
      },
      max_results: {
        type: 'number',
        description: 'Maximum number of results to return (1-10, default 5)',
      },
      search_depth: {
        type: 'string',
        enum: ['basic', 'advanced'],
        description: 'Search depth — "advanced" for more thorough results',
      },
      include_answer: {
        type: 'boolean',
        description: 'Whether to include an AI-generated answer summary (default true)',
      },
    },
    required: ['query'],
  },
  handler,
};

File: agent-service/tools/web_search.py
"""
Web search tool — Tavily API for current information.
"""

import httpx
from crewai.tools import BaseTool
from pydantic import BaseModel, Field
from config import TAVILY_API_KEY

TAVILY_URL = "https://api.tavily.com/search"


class WebSearchInput(BaseModel):
    query: str = Field(..., description="Search query")
    max_results: int = Field(5, description="Max results (1-10)", ge=1, le=10)
    search_depth: str = Field(
        "advanced", description='Search depth: "basic" or "advanced" (default advanced for better results)'
    )


class WebSearchTool(BaseTool):
    name: str = "web_search"
    description: str = (
        "Search the web for current information, news, trends, or external data. "
        "Use this when you need up-to-date information not in the knowledge base. "
        "Always prefer search_depth 'advanced' for thorough, high-quality results."
    )
    args_schema: type[BaseModel] = WebSearchInput

    def _run(
        self,
        query: str,
        max_results: int = 5,
        search_depth: str = "advanced",
    ) -> str:
        if not TAVILY_API_KEY:
            return "Web search is not configured (TAVILY_API_KEY missing)."

        query = query.strip()
        if not query:
            return "Missing search query."

        if search_depth not in ("basic", "advanced"):
            search_depth = "advanced"

        try:
            resp = httpx.post(
                TAVILY_URL,
                json={
                    "api_key": TAVILY_API_KEY,
                    "query": query,
                    "max_results": max_results,
                    "search_depth": search_depth,
                    "include_answer": True,
                    "include_raw_content": False,
                },
                timeout=30.0,
            )
            resp.raise_for_status()
            data = resp.json()

            output = ""
            if data.get("answer"):
                output += f"**Summary:** {data['answer']}\n\n"

            results = data.get("results", [])
            if results:
                output += "**Sources:**\n"
                for r in results:
                    snippet = (r.get("content") or "No snippet")[:300]
                    output += f"- [{r.get('title', 'Untitled')}]({r.get('url', '')})\n  {snippet}\n\n"
            else:
                output += "No results found for this query."

            return output.strip()

        except httpx.HTTPStatusError as e:
            return f"Tavily API error ({e.response.status_code}): {e.response.text[:200]}"
        except Exception as e:
            return f"Web search failed: {str(e)}"
